test/server/server_test.cc

```diff
@@ -197,10 +237,74 @@ class ServerInstanceImplTest : public testing::TestWithParam<Network::Address::I
   std::unique_ptr<InstanceImpl> server_;
 };
 
+// Custom StatsSink that just increments a counter when flush is called.
+class CustomStatsSink : public Stats::Sink {
+public:
+  CustomStatsSink(Stats::Scope& scope) : stats_flushed_(scope.counter("stats.flushed")) {}
+
+  // Stats::Sink
+  void flush(Stats::MetricSnapshot&) override { stats_flushed_.inc(); }
+
+  void onHistogramComplete(const Stats::Histogram&, uint64_t) override {}
+
+private:
+  Stats::Counter& stats_flushed_;
+};
+
+// Custom StatsSinFactory that creates CustomStatsSink.
+class CustomStatsSinkFactory : public Server::Configuration::StatsSinkFactory {
+public:
+  // StatsSinkFactory
+  Stats::SinkPtr createStatsSink(const Protobuf::Message&, Server::Instance& server) override {
+    return std::make_unique<CustomStatsSink>(server.stats());
+  }
+
+  ProtobufTypes::MessagePtr createEmptyConfigProto() override {
+    return ProtobufTypes::MessagePtr{new Envoy::ProtobufWkt::Empty()};
+  }
+
+  std::string name() override { return "envoy.custom_stats_sink"; }
+};
+
 INSTANTIATE_TEST_SUITE_P(IpVersions, ServerInstanceImplTest,
                          testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
                          TestUtility::ipTestParamsToString);
 
+/**
+ * Static registration for the custom sink factory. @see RegisterFactory.
+ */
+REGISTER_FACTORY(CustomStatsSinkFactory, Server::Configuration::StatsSinkFactory);
+
+// Validates that server stats are flushed even when server is stuck with initialization.
+TEST_P(ServerInstanceImplTest, StatsFlushWhenServerIsStillInitializing) {
+  absl::Notification started;
+
+  auto server_thread = Thread::threadFactoryForTest().createThread([&] {
```

Comment:
This is copied, can you factor this out into some type of `startupTestServer` function or something?

Reply:
done